File: sdks/go/suggest.go
package acli

// SuggestFlag returns a close known flag name or "" (spec §4.1).
func SuggestFlag(unknown string, known []string) string {
	best := ""
	bestDist := 3
	for _, k := range known {
		d := levenshtein(unknown, k)
		if d < bestDist {
			bestDist = d
			best = k
		}
	}
	return best
}

func levenshtein(a, b string) int {
	ra, rb := []rune(a), []rune(b)
	la, lb := len(ra), len(rb)
	if la == 0 {
		return lb
	}
	if lb == 0 {
		return la
	}
	m := make([][]int, la+1)
	for i := range m {
		m[i] = make([]int, lb+1)
		m[i][0] = i
	}
	for j := 0; j <= lb; j++ {
		m[0][j] = j
	}
	for i := 1; i <= la; i++ {
		for j := 1; j <= lb; j++ {
			cost := 0
			if ra[i-1] != rb[j-1] {
				cost = 1
			}
			m[i][j] = min3(m[i-1][j]+1, m[i][j-1]+1, m[i-1][j-1]+cost)
		}
	}
	return m[la][lb]
}

func min3(a, b, c int) int {
	if a <= b && a <= c {
		return a
	}
	if b <= c {
		return b
	}
	return c
}

File: sdks/dotnet/Acli.Spec/Suggest.cs
namespace Acli;

public static class SuggestFlag
{
    public static string? Suggest(string unknown, IEnumerable<string> known)
    {
        string? best = null;
        var bestDist = 3;
        foreach (var k in known)
        {
            var d = Levenshtein(unknown, k);
            if (d < bestDist)
            {
                bestDist = d;
                best = k;
            }
        }
        return best;
    }

    static int Levenshtein(string a, string b)
    {
        var la = a.Length;
        var lb = b.Length;
        var m = new int[la + 1, lb + 1];
        for (var i = 0; i <= la; i++) m[i, 0] = i;
        for (var j = 0; j <= lb; j++) m[0, j] = j;
        for (var i = 1; i <= la; i++)
        {
            for (var j = 1; j <= lb; j++)
            {
                var cost = a[i - 1] == b[j - 1] ? 0 : 1;
                m[i, j] = Math.Min(Math.Min(m[i - 1, j] + 1, m[i, j - 1] + 1), m[i - 1, j - 1] + cost);
            }
        }
        return m[la, lb];
    }
}
